cosmos-server/src/main/scala/com/mesosphere/cosmos/repository/UniverseClient.scala

```
@@ -21,3 +33,89 @@ object UniverseClient {
+
+  private[this] val stats = statsReceiver.scope("repositoryFetcher")
+
+  private[this] val cosmosVersion = {
+    val props = new Properties()
+    val is = this.getClass.getResourceAsStream("/build.properties")
+    if (is != null) {
+      props.load(is)
+      is.close()
+    }
+    Option(props.getProperty("cosmos.version")).getOrElse("unknown-version")
+  }
+
+  def apply(uri: Uri)(implicit session: RequestSession): Future[CosmosInternalRepository] = {
+    adminRouter.getDcosVersion().flatMap { dcosVersion =>
+      apply(uri, dcosVersion.version)
+    }
+  }
+
+  private[repository] def apply(uri: Uri, dcosReleaseVersion: DcosReleaseVersion)(implicit session: RequestSession): Future[CosmosInternalRepository] = {
+    Stat.timeFuture(stats.stat("fetch")) {
+      Future.value { uri.toURI.toURL.openConnection() }
+        .flatMap { conn =>
+          // Set headers on request
+          conn.setRequestProperty("Accept", MediaTypes.UniverseV3Repository.show)
+          conn.setRequestProperty("Accept-Encoding", "gzip")
+          conn.setRequestProperty("User-Agent", s"cosmos/$cosmosVersion dcos/${dcosReleaseVersion.show}")
+          session.authorization.foreach { auth =>
+            conn.setRequestProperty("Authorization", auth.headerValue)
+          }
+
+          // Handle the response
+          Future.value { Option(conn.getHeaderField("Content-Type")) -> Option(conn.getHeaderField("Content-Encoding")) }
+            .map { t =>
+              t._1.map(MediaTypeParser.parse(_).get) -> t._2
+            }
+            .map {
+              case (None, _)     => throw new IllegalStateException("Unable to process response without specified Content-Type")
+              case (Some(ct), ce) => ct -> ce
+            }
+            .map { case (contentType, contentEncoding) =>
+              val in = contentEncoding match {
+                case Some("gzip") => new GZIPInputStream(conn.getInputStream)
+                case _ => conn.getInputStream
```

Should we fail-fast here if the encoding has an unexpected value?